### PR TITLE
Refactoring to pull out code + flags that creates a database

### PR DESCRIPTION
**Pull request by Martin2112**
I'm expecting to pull more of the flag interactions code into the helper class in future PRs. I know the name server_helper isn't great but naming things is hard!

### CODE REVIEW

**Comment by pphaneuf** [comment]
Reviewed the first commit, moving on.

File: cpp/server/ct-mirror.cc [hunk 1]
```diff
@@ -331,40 +295,19 @@ int main(int argc, char* argv[]) {
...
 
-  const std::unique_ptr<EtcdClient> etcd_client(
-      stand_alone_mode
-          ? new FakeEtcdClient(event_base.get())
-          : new EtcdClient(&internal_pool, &url_fetcher,
-                           SplitHosts(FLAGS_etcd_servers)));
+  const unique_ptr<EtcdClient> etcd_client(
+      stand_alone_mode ? new FakeEtcdClient(event_base.get())
+                       : new EtcdClient(&internal_pool, &url_fetcher,
+                                        SplitHosts(FLAGS_etcd_servers)));
```
Comment (pphaneuf) [inline cpp/server/ct-mirror.cc]: Is this just whitespace reformatting? If so, grr...

File: cpp/server/ct-server.cc [hunk 2]
```diff
@@ -386,11 +323,10 @@ int main(int argc, char* argv[]) {
...
 
-  std::unique_ptr<EtcdClient> etcd_client(
-      stand_alone_mode
-          ? new FakeEtcdClient(event_base.get())
-          : new EtcdClient(&internal_pool, &url_fetcher,
-                           SplitHosts(FLAGS_etcd_servers)));
+  unique_ptr<EtcdClient> etcd_client(
+      stand_alone_mode ? new FakeEtcdClient(event_base.get())
+                       : new EtcdClient(&internal_pool, &url_fetcher,
+                                        SplitHosts(FLAGS_etcd_servers)));
```
Comment (pphaneuf) [inline cpp/server/ct-server.cc]: All this noise!

File: cpp/server/ct-mirror.cc (hunk 1 above)
Comment (Martin2112) [inline cpp/server/ct-mirror.cc]: No I removed a 'std::' and clang-format did the rest.

File: cpp/server/ct-server.cc (hunk 2 above)
Comment (Martin2112) [inline cpp/server/ct-server.cc]: OK that one is clang-format.

**Comment by pphaneuf** [comment]
I generally like the "unique_ptr for db reference" commit (other than the noise), but I can't help but think it'd be even better if `ProvideDatabase` returned a `unique_ptr<Database>`. In "modern C++", that's a common pattern to return a smart pointer, so that even if you screw up and ignore the return value, you don't leak.

**Comment by pphaneuf** [comment]
General comment: I find it nicer to have two blank lines between most top level things, à la Python. Look at `database.h` for an example.

**Comment by pphaneuf** [comment]
All right, went over all this, let me know when you've pushed some updates.

File: cpp/server/ct-mirror.cc (hunk 1 above)
Comment (pphaneuf) [inline cpp/server/ct-mirror.cc]: I approve of removing that `std::`, but I disapprove of not doing it separately. :wink:

File: cpp/server/ct-server.cc (hunk 2 above)
Comment (pphaneuf) [inline cpp/server/ct-server.cc]: There was an `std::` here too! Oh, how I'd love highlights *within* a diff line...

**Comment by Martin2112** [comment]
I'm not as keen on this because formatters don't support it and as I use an
IDE I can navigate around code easily without worrying about its fiddly
layout.

Plus most of our code doesn't have this so any alleged benefits we might
get from it don't currently exist anyway. And it's something done by Python
programmers (guilt by association). I rest my case m'lud.

**Comment by Martin2112** [comment]
ProvideDatabase() now returns a unique_ptr

**Comment by Martin2112** [comment]
Think it's mostly done. PTAL.

**Comment by pphaneuf** [comment]
LGTM, but I'm very curious to see what rebasing/squashing/merging strategy you'll take... :wink: 

image

**Comment by Martin2112** [comment]
I think I shall be using the "Friday Afternoon" strategy.

**Pull request merged**